swap/protocol_test.go

```diff
@@ -0,0 +1,317 @@
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	p2ptest "github.com/ethereum/go-ethereum/p2p/testing"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+/*
+TestHandshake creates two mock nodes and initiates an exchange;
+it expects a handshake to take place between the two nodes
+(the handshake would fail because we don't actually use real nodes here)
+*/
+func TestHandshake(t *testing.T) {
+	var err error
+
+	// setup test swap object
+	swap, dir := newTestSwap(t)
+	defer os.RemoveAll(dir)
+
+	ctx := context.Background()
+	testDeploy(ctx, swap.backend, swap)
+	// setup the protocolTester, which will allow protocol testing by sending messages
+	protocolTester := p2ptest.NewProtocolTester(swap.owner.privateKey, 2, swap.run)
+
+	// shortcut to creditor node
+	debitor := protocolTester.Nodes[0]
+	creditor := protocolTester.Nodes[1]
+
+	// set balance artifially
+	swap.balances[creditor.ID()] = -42
+
+	// create the expected cheque to be received
+	cheque := newTestCheque()
+
+	// sign the cheque
+	cheque.Signature, err = swap.signContent(cheque)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// run the exchange:
+	// trigger a `ChequeRequestMsg`
+	// expect a `EmitChequeMsg` with a valid cheque
+	err = protocolTester.TestExchanges(p2ptest.Exchange{
+		Label: "TestRequestCheque",
+		Triggers: []p2ptest.Trigger{
+			{
+				Code: 1,
+				Msg: &EmitChequeMsg{
+					Cheque: cheque,
+				},
+				Peer: debitor.ID(),
+			},
+		},
+		Expects: []p2ptest.Expect{
+			{
+				Code: 0,
+				Msg: &HandshakeMsg{
+					ContractAddress: swap.owner.Contract,
+				},
+				Peer: creditor.ID(),
+			},
+			{
+				Code: 0,
+				Msg: &HandshakeMsg{
+					ContractAddress: swap.owner.Contract,
+				},
+				Peer: debitor.ID(),
+			},
+		},
+	})
+
+	// there should be no error at this point
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestEmitCheque is a full round of a cheque exchange between peers via the protocol.
+// We create two swap, for the creditor (beneficiary) and debitor (issuer) each,
+// and deploy them to the simulated backend.
+// We then create Swap protocol peers with a MsgPipe to be able to directly write messages to each other.
+// We have the debitor send a cheque via an `EmitChequeMsg`, then the creditor "reads" (pipe) the message
+// and handles the cheque.
+func TestEmitCheque(t *testing.T) {
+	log.Debug("set up test swaps")
+	creditorSwap, testDir1 := newTestSwap(t)
+	debitorSwap, testDir2 := newTestSwap(t)
+	defer os.RemoveAll(testDir1)
+	defer os.RemoveAll(testDir2)
+
+	ctx := context.Background()
+
+	log.Debug("deploy to simulated backend")
+	testDeploy(ctx, creditorSwap.backend, creditorSwap)
+	testDeploy(ctx, debitorSwap.backend, debitorSwap)
+	creditorSwap.backend.(*backends.SimulatedBackend).Commit()
+	debitorSwap.backend.(*backends.SimulatedBackend).Commit()
+
+	log.Debug("create peer instances")
+	// create Peer instances
```

Comment:
clean up the comments here. i can't make much of this

Reply:
when we handle messages, a `Peer` object represents the remote peer. Thus, if I am the creditor, and I run `handleEmitMsg`, then `Peer` in that function is an object of the remote peer - thus the debitor.

As we are creating peers manually for this test, these comments make clear which object is what. Thus `creditor` is the `Peer` object when running messages inside the debitor node, and vice-versa. I think these comments are extremely important to understand what the test is doing. I insist we should keep them.